server/src/test/java/org/elasticsearch/cluster/routing/allocation/RoutingNodesIntegrityTests.java

```diff
@@ -142,59 +142,41 @@ public void testBalanceIncrementallyStartNodes() {
             org.elasticsearch.cluster.ClusterName.CLUSTER_NAME_SETTING.getDefault(Settings.EMPTY)
         ).metadata(metadata).routingTable(initialRoutingTable).build();
 
-        logger.info("Adding one node and performing rerouting");
+        logger.info("Adding node-1 and performing reroute");
         clusterState = ClusterState.builder(clusterState).nodes(DiscoveryNodes.builder().add(newNode("node1"))).build();
-
         clusterState = strategy.reroute(clusterState, "reroute");
 
-        logger.info("Add another node and perform rerouting, nothing will happen since primary not started");
+        logger.info("Add node-2 and perform reroute, nothing will happen since primary not started");
         clusterState = ClusterState.builder(clusterState).nodes(DiscoveryNodes.builder(clusterState.nodes()).add(newNode("node2"))).build();
         clusterState = strategy.reroute(clusterState, "reroute");
 
-        logger.info("Start the primary shard");
-        clusterState = startInitializingShardsAndReroute(strategy, clusterState);
-
-        logger.info("Reroute, nothing should change");
-        clusterState = strategy.reroute(clusterState, "reroute");
-
-        logger.info("Start the backup shard");
-        clusterState = startInitializingShardsAndReroute(strategy, clusterState);
+        logger.info("Start the all shards");
+        clusterState = startInitializingShardsAndReroute(strategy, clusterState); // primaries
+        clusterState = startInitializingShardsAndReroute(strategy, clusterState); // replicas
 
-        clusterState = startInitializingShardsAndReroute(strategy, clusterState);
-
-        logger.info("Add another node and perform rerouting, relocate shards to new node");
+        logger.info("Add node-3 and perform reroute, relocate shards to new node");
         clusterState = ClusterState.builder(clusterState).nodes(DiscoveryNodes.builder(clusterState.nodes()).add(newNode("node3"))).build();
         clusterState = strategy.reroute(clusterState, "reroute");
 
-        logger.info("Reroute, nothing should change");
-        ClusterState newState = strategy.reroute(clusterState, "reroute");
-        assertThat(newState, equalTo(clusterState));
-
-        logger.info("Start the backup shard");
-        newState = startInitializingShardsAndReroute(strategy, clusterState);
-        assertThat(newState, not(equalTo(clusterState)));
-        clusterState = newState;
+        logger.info("Await all shards reallocate");
+        do {
+            clusterState = startInitializingShardsAndReroute(strategy, clusterState);
+        } while (shardsWithState(clusterState.getRoutingNodes(), INITIALIZING).size() > 0);
```

Comment:
Could we just call `applyStartedShardsUntilNoChange`? It's slightly different (no change vs no INITIALIZING shards) but maybe equivalent here?